src/launcher/java/org/truffleruby/launcher/RubyLauncher.java

```diff
@@ -488,4 +446,47 @@ private static void printShortHelp(PrintStream out) {
         out.println("  -h              show this message, --help for more info");
     }
 
+    public static class Metrics {
+
+        // These system properties are used before outside the SDK option system
+        public static boolean METRICS_TIME;
+        private static final boolean METRICS_MEMORY_USED_ON_EXIT =
+                Boolean.getBoolean("truffleruby.metrics.memory_used_on_exit");
+
+        public static void printTruffleTime(String id) {
+            if (METRICS_TIME) {
+                final long millis = System.currentTimeMillis();
+                System.err.printf("%s %d.%03d%n", id, millis / 1000, millis % 1000);
+            }
+        }
+
+        private static void printTruffleMemory() {
+            // Memory stats aren't available in native.
+            if (!isAOT() && METRICS_MEMORY_USED_ON_EXIT) {
+                for (int n = 0; n < 10; n++) {
+                    System.gc();
+                }
+
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+
+                System.err.printf("allocated %d%n", ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getUsed());
+            }
+        }
+
+        static void end() {
+            printTruffleTime("after-main");
+            printTruffleMemory();
+        }
+
+        static void begin() {
+            // Assigned here so it's available on SVM as well
+            METRICS_TIME = Boolean.getBoolean("truffleruby.metrics.time");
+
+            printTruffleTime("before-main");
+        }
+    }
```

Comment:
Looks nice, but please extract it to a top-level class (in its own file).
The metrics are not really related to the launcher, they just happen to be used by it.
This will also avoid the very verbose `RubyLauncher.Metrics.printTruffleTime` below.

Reply:
isAot is protected, it would have to be exposed under different name, therefore it seemed better option to just leave it in launcher until isAot is made public.

Reply:
I'll add a todo.